llm/openai/openai.go

```diff
@@ -206,6 +206,40 @@ func (o *OpenAI) Generate(ctx context.Context, t *thread.Thread) error {
 	return nil
 }
 
+func isStreamToolCallResponse(response *openai.ChatCompletionStreamResponse) bool {
+	return response.Choices[0].FinishReason == openai.FinishReasonToolCalls || len(response.Choices[0].Delta.ToolCalls) > 0
+}
+
+func consumeToolCallResponse(response *openai.ChatCompletionStreamResponse,
+	currentToolCall *openai.ToolCall) (updatedToolCall openai.ToolCall, isNewTool bool) {
+	if len(response.Choices[0].Delta.ToolCalls) == 0 {
+		return
+	}
+	if response.Choices[0].Delta.ToolCalls[0].ID != "" {
+		isNewTool = true
+		updatedToolCall = response.Choices[0].Delta.ToolCalls[0]
+	} else {
+		currentToolCall.Function.Arguments += response.Choices[0].Delta.ToolCalls[0].Function.Arguments
+	}
+	return
+}
+
+func (o *OpenAI) processErrorResponse(messages []*thread.Message, content string,
+	currentToolCall *openai.ToolCall, allToolCalls []openai.ToolCall) []*thread.Message {
```

Comment:
Let's call it `handleEndOfStream` 'cause even though `EOF` is an error it means that there is no more data on the stream.

```suggestion
func (o *OpenAI) handleEndOfStream(
	messages []*thread.Message,
	content string,
	currentToolCall *openai.ToolCall,
	allToolCalls []openai.ToolCall,
) []*thread.Message {
```